### ISSUE TITLE
Change in runif() behavior in Rcpp 0.12.16

### ISSUE THREAD

**Issue by kbroman** [RcppCore/Rcpp#836]
With Rcpp 0.12.16, my [simcross](https://github.com/kbroman/simcross) package started throwing errors  in the tests. I don't fully understand what's going on, but it seems related to me calling `runif(n,0,L)` with `n==0`.

Consider the following code (also [available as a gist](https://gist.github.com/kbroman/aedc8d8313f5f471e7089c3d0205f45e)). This simulates a Poisson process on the interval (0,L) by drawing a Poisson number of points and then drawing their positions from a uniform distribution. There are two versions of the function; in the first, I don't try to trap the `n==0` cases; in the latter I call `runif()` only with `n>0`.

```cpp
#include <Rcpp.h>
using namespace Rcpp;

// [[Rcpp::export]]
NumericVector sim_crossovers_orig(const double L)
{
    int n_xo;

    n_xo = R::rpois(L/100.0);

    NumericVector tmp = runif(n_xo, 0.0, L);
    return tmp.sort();
}

// [[Rcpp::export]]
NumericVector sim_crossovers_new(const double L)
{
    int n_xo;

    n_xo = R::rpois(L/100.0);

    NumericVector tmp(0);
    if(n_xo > 0) tmp = runif(n_xo, 0.0, L);
    return tmp.sort();
}
```

With Rcpp ver 0.12.15, these two functions give the same results.

```
Rcpp::sourceCpp("sim_crossovers.cpp")

set.seed(20180318)
z <- replicate(10000, sim_crossovers_orig(100))
table(sapply(z, length))
##    0    1    2    3    4    5    6
## 3661 3681 1837  631  151   35    4

set.seed(20180318)
z <- replicate(10000, sim_crossovers_new(100))
table(sapply(z, length))
##    0    1    2    3    4    5    6
## 3661 3681 1837  631  151   35    4
```

But with Rcpp ver 0.12.16, the former gives a ton of 0-length vectors, while the latter works fine, though it gives somewhat different results from ver 0.12.15. (The tables being printed should be 10,000 draws from a Poisson distribution with mean 1.)

```
Rcpp::sourceCpp("sim_crossovers.cpp")

set.seed(20180318)
z <- replicate(10000, sim_crossovers_orig(100))
table(sapply(z, length))
##    0    1    3
## 9994    5    1

set.seed(20180318)
z <- replicate(10000, sim_crossovers_new(100))
table(sapply(z, length))
##    0    1    2    3    4    5    6
## 3637 3698 1846  621  152   41    5
```

I've tried this in both Linux and MacOS (High Sierra), and with R 3.4.3 and 3.4.4, with the same results in all cases. (Just the version of Rcpp matters.)

Here's my session info in the Rcpp 0.12.16 case on my Mac:

```
R version 3.4.4 (2018-03-15)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS High Sierra 10.13.3

Matrix products: default
BLAS: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] Rcpp_0.12.16    devtools_1.13.5

loaded via a namespace (and not attached):
[1] compiler_3.4.4   tools_3.4.4      withr_2.1.1.9000 memoise_1.1.0    digest_0.6.15
```

**Comment by kevinushey** [comment]
This definitely has something to do with the RNG state. I suspect I broke this somehow with this commit:

https://github.com/RcppCore/Rcpp/commit/573e71df4976e8e7d51e6dd5b0d7b4e7b1478665

The following example provides the expected results:

```cpp
#include <Rcpp.h>
using namespace Rcpp;

// [[Rcpp::export]]
NumericVector sim_crossovers_orig(const double L)
{
   int n_xo;

   GetRNGstate();
   n_xo = R::rpois(L/100.0);
   PutRNGstate();

   GetRNGstate();
   NumericVector tmp = runif(n_xo, 0.0, L);
   PutRNGstate();

   return tmp.sort();
}

/*** R
set.seed(20180318)
z <- replicate(10000, sim_crossovers_orig(100))
table(sapply(z, length))
##    0    1    2    3    4    5    6
## 3661 3681 1837  631  151   35    4
*/
```

**Comment by kevinushey** [comment]
Argh. Let's understand where this issue is coming from. The call to `runif()` is implemented as:

https://github.com/RcppCore/Rcpp/blob/9c1c356e9f73f02f7121c2e07d16080c4c1acbde/inst/include/Rcpp/stats/random/random.h#L307-L309

We use the 'fill' constructor of `NumericVector`, passing it a generator that generated random uniforms. That 'generator' thing is defined here:

https://github.com/RcppCore/Rcpp/blob/9c1c356e9f73f02f7121c2e07d16080c4c1acbde/inst/include/Rcpp/stats/random/runif.h#L45-L55

It inherits from something called 'Generator'. What is that? Well, it lives here:

https://github.com/RcppCore/Rcpp/blob/9c1c356e9f73f02f7121c2e07d16080c4c1acbde/inst/include/Rcpp/stats/random/random.h#L27-L31

In other words, all our generator objects _inherit_ from RNGScope. Which means every time a generator is constructed, and then destructed, it's now (indiscriminately) getting and setting the RNG state. Before, this didn't happen because these RNGScope calls were effectively screened by a global counter.

tl;dr: I broke it. Sorry. :-/

The issue can be fixed by making sure our generators don't inherit from RNGScope.

**Comment by eddelbuettel** [comment]
Nice analysis.  The only other breakage we had seen was in ECctmc where I liased with the maintainer.  So I guess the breakage was a little wider that we thought :-/

**Comment by kbroman** [comment]
thanks, @kevinushey!

**Comment by eddelbuettel** [comment]
@kbroman For completeness, were you able to check that the now-merged PR fixes your test failures?

**Comment by kbroman** [comment]
@eddelbuettel Yes, the PR fixed my test failures; thanks!

**Comment by eddelbuettel** [comment]
Thanks for confirming, and sorry for the breakage.  We usually manage to do better, but the shiny folks wanted a change so ... it happened.

**Comment by kbroman** [comment]
@eddelbuettel Absolutely no problem; only praise from me for Rcpp! And reducing my mess to a reasonably concise reproducible example was an interesting puzzle.
